File: javascript/spiral_matrix.js
let mtrx = [[1, 2, 3, 4],
[4, 5, 6, 7],
[7, 8, 9, 10],
[10,11,12,13]];

let res = [];

let top = 0;
let bottom = mtrx.length;
let left = 0;
let right = mtrx[0].length;

while (top < bottom && left < right) {
    for (let i = left; i < right; i++) {
        res.push(mtrx[top][i]);
    }
    top += 1;

    for (let i = top; i < bottom; i++) {
        res.push(mtrx[i][right - 1]);
    }
    right -= 1;

    if (top < bottom) {

        for (let i = right - 1; i >= left; i--) {
            res.push(mtrx[bottom - 1][i]);
        }
        bottom -= 1;

    }

    if (left < right) {
        for (let i = bottom - 1; i >= top; i--) {
            res.push(mtrx[i][left]);
        }
        left += 1;
    }
}

console.log(res)

File: java/SpiralMatrix.java
// https://leetcode.com/problems/spiral-matrix-ii/
import java.util.*;

class Solution {
    public int[][] generateMatrix(int n) {
        int[][] matrix = new int[n][n];
        
        int left = 0, right = n - 1;
        int top = 0, bottom = n - 1;
        int num = 1;
        
        while (left <= right && top <= bottom) {
            // fill top row
            for (int i = left; i <= right; i++) {
                matrix[top][i] = num++;
            }
            top++;
            
            // fill right column
            for (int i = top; i <= bottom; i++) {
                matrix[i][right] = num++;
            }
            right--;
            
            // fill bottom row
            if (top <= bottom) {
                for (int i = right; i >= left; i--) {
                    matrix[bottom][i] = num++;
                }
                bottom--;
            }
            
            // fill left column
            if (left <= right) {
                for (int i = bottom; i >= top; i--) {
                    matrix[i][left] = num++;
                }
                left++;
            }
        }
        
        return matrix;
    }
}

// Time Complexity: O(n^2)
// Space Complexity: O(1) (excluding the output matrix)